File: src/models/Workspace.js
const db = require('./database');

class Workspace {
  static getAll(filters = {}) {
    let query = 'SELECT * FROM workspaces WHERE 1=1';
    const params = [];

    if (filters.user_name) {
      query += ' AND user_name LIKE ?';
      params.push(`%${filters.user_name}%`);
    }

    if (filters.state) {
      query += ' AND state = ?';
      params.push(filters.state);
    }

    if (filters.bundle_id) {
      query += ' AND bundle_id = ?';
      params.push(filters.bundle_id);
    }

    if (filters.running_mode) {
      query += ' AND running_mode = ?';
      params.push(filters.running_mode);
    }

    if (filters.created_from) {
      query += ' AND created_at >= ?';
      params.push(filters.created_from);
    }

    if (filters.created_to) {
      query += ' AND created_at <= ?';
      params.push(filters.created_to);
    }

    if (filters.terminated) {
      if (filters.terminated === 'true') {
        query += ' AND terminated_at IS NOT NULL';
      } else {
        query += ' AND terminated_at IS NULL';
      }
    }

    query += ' ORDER BY created_at DESC';

    if (filters.limit) {
      query += ' LIMIT ?';
      params.push(parseInt(filters.limit, 10));
    }

    if (filters.offset) {
      query += ' OFFSET ?';
      params.push(parseInt(filters.offset, 10));
    }

    return db.prepare(query).all(...params);
  }

  static getById(id) {
    return db.prepare('SELECT * FROM workspaces WHERE id = ?').get(id);
  }

  static upsert(workspace) {
    const stmt = db.prepare(`
      INSERT INTO workspaces (
        id, directory_id, user_name, ip_address, state, bundle_id,
        subnet_id, computer_name, running_mode, running_mode_auto_stop_timeout_in_minutes,
        root_volume_size_gib, user_volume_size_gib, created_at, terminated_at,
        last_known_user_connection_timestamp, tags, updated_at
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, CURRENT_TIMESTAMP)
      ON CONFLICT(id) DO UPDATE SET
        directory_id = excluded.directory_id,
        user_name = excluded.user_name,
        ip_address = excluded.ip_address,
        state = excluded.state,
        bundle_id = excluded.bundle_id,
        subnet_id = excluded.subnet_id,
        computer_name = excluded.computer_name,
        running_mode = excluded.running_mode,
        running_mode_auto_stop_timeout_in_minutes = excluded.running_mode_auto_stop_timeout_in_minutes,
        root_volume_size_gib = excluded.root_volume_size_gib,
        user_volume_size_gib = excluded.user_volume_size_gib,
        terminated_at = excluded.terminated_at,
        last_known_user_connection_timestamp = excluded.last_known_user_connection_timestamp,
        tags = excluded.tags,
        updated_at = CURRENT_TIMESTAMP
    `);

    return stmt.run(
      workspace.id,
      workspace.directory_id,
      workspace.user_name,
      workspace.ip_address,
      workspace.state,
      workspace.bundle_id,
      workspace.subnet_id,
      workspace.computer_name,
      workspace.running_mode,
      workspace.running_mode_auto_stop_timeout_in_minutes,
      workspace.root_volume_size_gib,
      workspace.user_volume_size_gib,
      workspace.created_at,
      workspace.terminated_at,
      workspace.last_known_user_connection_timestamp,
      JSON.stringify(workspace.tags || {})
    );
  }

  static getCount(filters = {}) {
    let query = 'SELECT COUNT(*) as count FROM workspaces WHERE 1=1';
    const params = [];

    if (filters.user_name) {
      query += ' AND user_name LIKE ?';
      params.push(`%${filters.user_name}%`);
    }

    if (filters.state) {
      query += ' AND state = ?';
      params.push(filters.state);
    }

    if (filters.bundle_id) {
      query += ' AND bundle_id = ?';
      params.push(filters.bundle_id);
    }

    if (filters.running_mode) {
      query += ' AND running_mode = ?';
      params.push(filters.running_mode);
    }

    return db.prepare(query).get(...params);
  }

  static getStates() {
    return db.prepare('SELECT DISTINCT state FROM workspaces WHERE state IS NOT NULL').all();
  }

  static getBundles() {
    return db.prepare('SELECT DISTINCT bundle_id FROM workspaces WHERE bundle_id IS NOT NULL').all();
  }

  static getRunningModes() {
    return db.prepare('SELECT DISTINCT running_mode FROM workspaces WHERE running_mode IS NOT NULL').all();
  }
}

module.exports = Workspace;

File: backend/models/workspace.go
package models

import (
	"database/sql"
	"encoding/json"
	"time"
)

// Workspace represents an AWS WorkSpace
type Workspace struct {
	WorkspaceID                         string          `json:"workspace_id" db:"workspace_id"`
	UserName                            string          `json:"user_name" db:"user_name"`
	DisplayName                         string          `json:"display_name" db:"display_name"`
	DirectoryID                         string          `json:"directory_id" db:"directory_id"`
	IPAddress                           string          `json:"ip_address" db:"ip_address"`
	State                               string          `json:"state" db:"state"`
	BundleID                            string          `json:"bundle_id" db:"bundle_id"`
	SubnetID                            string          `json:"subnet_id" db:"subnet_id"`
	ComputerName                        string          `json:"computer_name" db:"computer_name"`
	RunningMode                         string          `json:"running_mode" db:"running_mode"`
	RootVolumeSizeGib                   int             `json:"root_volume_size_gib" db:"root_volume_size_gib"`
	UserVolumeSizeGib                   int             `json:"user_volume_size_gib" db:"user_volume_size_gib"`
	ComputeTypeName                     string          `json:"compute_type_name" db:"compute_type_name"`
	CreatedAt                           *time.Time      `json:"created_at" db:"created_at"`
	TerminatedAt                        *time.Time      `json:"terminated_at" db:"terminated_at"`
	LastKnownUserConnectionTimestamp    *time.Time      `json:"last_known_user_connection_timestamp" db:"last_known_user_connection_timestamp"`
	CreatedByUser                       string          `json:"created_by_user" db:"created_by_user"`
	TerminatedByUser                    string          `json:"terminated_by_user" db:"terminated_by_user"`
	Tags                                json.RawMessage `json:"tags" db:"tags"`
	UpdatedAt                           time.Time       `json:"updated_at" db:"updated_at"`
}

// WorkspaceUsage represents monthly usage data for a workspace
type WorkspaceUsage struct {
	ID          int       `json:"id" db:"id"`
	WorkspaceID string    `json:"workspace_id" db:"workspace_id"`
	Month       string    `json:"month" db:"month"` // YYYY-MM format
	UsageHours  float64   `json:"usage_hours" db:"usage_hours"`
	CreatedAt   time.Time `json:"created_at" db:"created_at"`
	UpdatedAt   time.Time `json:"updated_at" db:"updated_at"`
}

// GetWorkspaceByID retrieves a workspace by ID
func GetWorkspaceByID(db *sql.DB, workspaceID string) (*Workspace, error) {
	var ws Workspace
	query := `
		SELECT workspace_id, user_name, display_name, directory_id, ip_address,
		       state, bundle_id, subnet_id, computer_name, running_mode,
		       root_volume_size_gib, user_volume_size_gib, compute_type_name,
		       created_at, terminated_at, last_known_user_connection_timestamp,
		       created_by_user, terminated_by_user, tags, updated_at
		FROM workspaces
		WHERE workspace_id = $1
	`
	err := db.QueryRow(query, workspaceID).Scan(
		&ws.WorkspaceID, &ws.UserName, &ws.DisplayName, &ws.DirectoryID, &ws.IPAddress,
		&ws.State, &ws.BundleID, &ws.SubnetID, &ws.ComputerName, &ws.RunningMode,
		&ws.RootVolumeSizeGib, &ws.UserVolumeSizeGib, &ws.ComputeTypeName,
		&ws.CreatedAt, &ws.TerminatedAt, &ws.LastKnownUserConnectionTimestamp,
		&ws.CreatedByUser, &ws.TerminatedByUser, &ws.Tags, &ws.UpdatedAt,
	)
	if err != nil {
		return nil, err
	}
	return &ws, nil
}

// ListWorkspaces retrieves workspaces with filtering and pagination
func ListWorkspaces(db *sql.DB, filters map[string]interface{}, limit, offset int) ([]Workspace, int, error) {
	// Build query with filters
	baseQuery := `
		SELECT workspace_id, user_name, display_name, directory_id, ip_address,
		       state, bundle_id, subnet_id, computer_name, running_mode,
		       root_volume_size_gib, user_volume_size_gib, compute_type_name,
		       created_at, terminated_at, last_known_user_connection_timestamp,
		       created_by_user, terminated_by_user, tags, updated_at
		FROM workspaces
		WHERE 1=1
	`

	countQuery := "SELECT COUNT(*) FROM workspaces WHERE 1=1"
	args := []interface{}{}
	argPos := 1

	// Apply filters
	if userName, ok := filters["user_name"].(string); ok && userName != "" {
		baseQuery += " AND user_name = $" + string(rune(argPos+'0'))
		countQuery += " AND user_name = $" + string(rune(argPos+'0'))
		args = append(args, userName)
		argPos++
	}

	if state, ok := filters["state"].(string); ok && state != "" {
		baseQuery += " AND state = $" + string(rune(argPos+'0'))
		countQuery += " AND state = $" + string(rune(argPos+'0'))
		args = append(args, state)
		argPos++
	}

	// Get total count
	var total int
	err := db.QueryRow(countQuery, args...).Scan(&total)
	if err != nil {
		return nil, 0, err
	}

	// Add pagination
	baseQuery += " ORDER BY created_at DESC LIMIT $" + string(rune(argPos+'0')) + " OFFSET $" + string(rune(argPos+1+'0'))
	args = append(args, limit, offset)

	rows, err := db.Query(baseQuery, args...)
	if err != nil {
		return nil, 0, err
	}
	defer rows.Close()

	workspaces := []Workspace{}
	for rows.Next() {
		var ws Workspace
		err := rows.Scan(
			&ws.WorkspaceID, &ws.UserName, &ws.DisplayName, &ws.DirectoryID, &ws.IPAddress,
			&ws.State, &ws.BundleID, &ws.SubnetID, &ws.ComputerName, &ws.RunningMode,
			&ws.RootVolumeSizeGib, &ws.UserVolumeSizeGib, &ws.ComputeTypeName,
			&ws.CreatedAt, &ws.TerminatedAt, &ws.LastKnownUserConnectionTimestamp,
			&ws.CreatedByUser, &ws.TerminatedByUser, &ws.Tags, &ws.UpdatedAt,
		)
		if err != nil {
			return nil, 0, err
		}
		workspaces = append(workspaces, ws)
	}

	return workspaces, total, nil
}

// UpsertWorkspace inserts or updates a workspace
func UpsertWorkspace(db *sql.DB, ws *Workspace) error {
	query := `
		INSERT INTO workspaces (
			workspace_id, user_name, display_name, directory_id, ip_address,
			state, bundle_id, subnet_id, computer_name, running_mode,
			root_volume_size_gib, user_volume_size_gib, compute_type_name,
			created_at, terminated_at, last_known_user_connection_timestamp,
			created_by_user, terminated_by_user, tags, updated_at
		) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, CURRENT_TIMESTAMP)
		ON CONFLICT (workspace_id) DO UPDATE SET
			user_name = EXCLUDED.user_name,
			display_name = EXCLUDED.display_name,
			directory_id = EXCLUDED.directory_id,
			ip_address = EXCLUDED.ip_address,
			state = EXCLUDED.state,
			bundle_id = EXCLUDED.bundle_id,
			subnet_id = EXCLUDED.subnet_id,
			computer_name = EXCLUDED.computer_name,
			running_mode = EXCLUDED.running_mode,
			root_volume_size_gib = EXCLUDED.root_volume_size_gib,
			user_volume_size_gib = EXCLUDED.user_volume_size_gib,
			compute_type_name = EXCLUDED.compute_type_name,
			created_at = EXCLUDED.created_at,
			terminated_at = EXCLUDED.terminated_at,
			last_known_user_connection_timestamp = EXCLUDED.last_known_user_connection_timestamp,
			created_by_user = EXCLUDED.created_by_user,
			terminated_by_user = EXCLUDED.terminated_by_user,
			tags = EXCLUDED.tags,
			updated_at = CURRENT_TIMESTAMP
	`

	_, err := db.Exec(query,
		ws.WorkspaceID, ws.UserName, ws.DisplayName, ws.DirectoryID, ws.IPAddress,
		ws.State, ws.BundleID, ws.SubnetID, ws.ComputerName, ws.RunningMode,
		ws.RootVolumeSizeGib, ws.UserVolumeSizeGib, ws.ComputeTypeName,
		ws.CreatedAt, ws.TerminatedAt, ws.LastKnownUserConnectionTimestamp,
		ws.CreatedByUser, ws.TerminatedByUser, ws.Tags,
	)

	return err
}
